From 1431e0ec6c85c984a2233acb4b3e5c345701604a Mon Sep 17 00:00:00 2001
From: Stephan Hartmann <stha09@googlemail.com>
Date: Mon, 25 Apr 2022 12:07:42 +0000
Subject: [PATCH] libstdc++: fix incomplete type of media::MediaLog

Destructor of std::unique_ptr in libstdc++ uses sizeof() which
requires full definition of media::MediaLog in blink::Platform.
Move definition to platform.cc to avoid large header inclusion.

Bug: 957519
Change-Id: If060fa83ee37816ef428707d30c711de36316c64
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/3605223
Reviewed-by: Kentaro Hara <haraken@chromium.org>
Commit-Queue: Stephan Hartmann <stha09@googlemail.com>
Cr-Commit-Position: refs/heads/main@{#995659}
---
 third_party/blink/public/platform/platform.h             | 4 +---
 third_party/blink/renderer/platform/exported/DEPS        | 1 +
 third_party/blink/renderer/platform/exported/platform.cc | 8 ++++++++
 3 files changed, 10 insertions(+), 3 deletions(-)

diff --git a/third_party/blink/public/platform/platform.h b/third_party/blink/public/platform/platform.h
index 3a8c825cbbc19..3f8fbe6330051 100644
--- a/third_party/blink/public/platform/platform.h
+++ b/third_party/blink/public/platform/platform.h
@@ -811,9 +811,7 @@ class BLINK_PLATFORM_EXPORT Platform {
   virtual std::unique_ptr<media::MediaLog> GetMediaLog(
       MediaInspectorContext* inspector_context,
       scoped_refptr<base::SingleThreadTaskRunner> owner_task_runner,
-      bool is_on_worker) {
-    return nullptr;
-  }
+      bool is_on_worker);
 
   // GpuVideoAcceleratorFactories --------------------------------------
 
diff --git a/third_party/blink/renderer/platform/exported/DEPS b/third_party/blink/renderer/platform/exported/DEPS
index feaedc79fdcb1..b73dbd34f0f3b 100644
--- a/third_party/blink/renderer/platform/exported/DEPS
+++ b/third_party/blink/renderer/platform/exported/DEPS
@@ -1,4 +1,5 @@
 include_rules = [
+    "+media/base/media_log.h",
     "+net/base/ip_endpoint.h",
     "+net/base/load_flags.h",
     "+net/ssl/ssl_info.h",
diff --git a/third_party/blink/renderer/platform/exported/platform.cc b/third_party/blink/renderer/platform/exported/platform.cc
index 496596f8a4b1a..3ecc1f3bc2e4a 100644
--- a/third_party/blink/renderer/platform/exported/platform.cc
+++ b/third_party/blink/renderer/platform/exported/platform.cc
@@ -40,6 +40,7 @@
 #include "build/build_config.h"
 #include "components/viz/common/gpu/raster_context_provider.h"
 #include "gpu/ipc/client/gpu_channel_host.h"
+#include "media/base/media_log.h"
 #include "services/network/public/cpp/shared_url_loader_factory.h"
 #include "third_party/blink/public/common/thread_safe_browser_interface_broker_proxy.h"
 #include "third_party/blink/public/platform/scheduler/web_thread_scheduler.h"
@@ -381,4 +382,11 @@ gfx::ColorSpace Platform::GetRenderingColorSpace() const {
   return {};
 }
 
+std::unique_ptr<media::MediaLog> Platform::GetMediaLog(
+    MediaInspectorContext* inspector_context,
+    scoped_refptr<base::SingleThreadTaskRunner> owner_task_runner,
+    bool is_on_worker) {
+  return nullptr;
+}
+
 }  // namespace blink
